pkg/cloudprovider/aws/instancetype.go

```diff
@@ -270,7 +269,7 @@ func (i *InstanceType) computeOverhead(vmMemOverhead float64) v1.ResourceList {
 			// vm-overhead
 			(int64(math.Ceil(float64(memory.Value())*vmMemOverhead/1024/1024)))+
 				// kube-reserved
-				((11*i.eniLimitedPods())+255)+
+				((11*pods.Value())+255)+
```

Comment:
Interesting! Bottlerocket seems to behave differently then, as you can see [here](https://github.com/bottlerocket-os/bottlerocket/blob/5d69ebc1fc76a6032b677c9f3941bfa58221f3af/packages/kubernetes-1.22/kubelet-config#L71) and [here](https://github.com/bottlerocket-os/bottlerocket/blob/862c4fa3b9c09a8f6696c360a63ba644b3caaf35/sources/api/schnauzer/src/helpers.rs#L921). When max-pods is set in the user data it seems to always configure `max_pods*11+255 Mi` for the kube reserved memory. Without it, it will take the ENI based pod limits (or 110 as default if it doesn't have the data, e.g. for new instance types).

While I agree that this is a waste on small instance types, I would argue that Karpenter should calculate in the same ways as the AMIs do. Maybe a future update could consider a more dynamic max pods setting for non ENI based pod density (e.g. scale based on available vCPUs)? The code from this PR should already work out of the box then.

I do wonder what an appropriate way forward would be. Since these values are pre-calculated I don't think I can build a switch statement based on the AMI family here. I can add some code to the AL2 AMI user data generation that will override the reserved memory with the more appropriate one maybe? That would have the additional benefit of working around the issue you mentioned. What do you think?

Reply:
Ah I see - so the reason you're seeing a difference in behavior is because you're using Bottlerocket and kubeReserved is being calculated differently by that AMI (compared to AL2). I see that your difference in allocatable as per #1959 was `84.60 Mebibytes` (`9037933121 - 8949215232` bytes). But if I'm calculating correctly, I'd have expected a larger diff since 893Mb is what we're calculating (11 * 58 + 255) in the binpacking logic,  and the BottleRocket AMI is doing 1465Mb.

I agree that we should match the calculation of the AMI in the bin-packing logic and I'm open to changing the way we bin-pack for Bottlerocket nodes. 

In th elong term, I think we want the maxPods memory calculation to be a function of the memory on the instanceType and use something [like this](https://cloud.google.com/kubernetes-engine/docs/concepts/cluster-architecture#memory_cpu).

Reply:
I just had another look with a fresh set of eyes and realized that my thoughts from the last comment were way off. I'll implement the calculation adjustment based on AMI family as you suggested and will report back. Thanks for all your input!